File: backend/backend/sync_texts.py
import argparse
import os
from pathlib import Path
from backend.utils.utils_supabase import init_supabase
import markdown
from loguru import logger

log = logger

def get_all_markdown_files(dir_path: Path, base: Path = Path()) -> list[str]:
    files = []
    for entry in dir_path.iterdir():
        rel_path = base / entry.name
        if entry.is_dir():
            files.extend(get_all_markdown_files(entry, rel_path))
        elif entry.is_file() and entry.suffix == '.md':
            files.append(str(rel_path))
    return files

def markdown_to_html(md_text: str) -> str:
    return markdown.markdown(md_text)

def sync_texts(env_file_path: str, markdown_files_path: str):
    BASE_DIR = Path(markdown_files_path)
    markdown_files = get_all_markdown_files(BASE_DIR)

    if not markdown_files:
        log.info(f"⚠️  No Markdown files found in {BASE_DIR}. Nothing to upload.")
        return

    supabase = init_supabase(env_file_path)

    supabase.table('phase_texts').delete().neq('id', 0).execute()

    for rel_path in markdown_files:
        full_path = BASE_DIR / rel_path
        with open(full_path, 'r', encoding='utf-8') as f:
            md_content = f.read()
        html_content = markdown_to_html(md_content)

        try:
            response = supabase.table('phase_texts').upsert({
                'path': rel_path,
                'html_content': html_content
            }).execute()
            log.info(f"✅ Uploaded {rel_path} ({len(response.data) if response.data else 0} row(s))")
        except Exception as e:
            log.error(f"❌ Failed to upload {rel_path}: {e}")

    log.info("✅ All markdown files uploaded to Supabase")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description='Script to sync phase texts to the supabase database')

    parser.add_argument(
        '--env_file',
        help='Path to the .env file',
        type=str,
        default=None,
        dest='env_file_path',
        required=True,
    )

    parser.add_argument(
        '--md_path',
        help='Path to the markdown folder',
        type=str,
        default=None,
        dest='markdown_files_path',
        required=True,
    )
    return parser.parse_args() 

if __name__ == "__main__":
    args = parse_args()
    sync_texts(**vars(args))

File: frontend/scripts/sync-texts.ts
import { Database } from "@/types/database.types";
import { SupabaseClient } from "@supabase/supabase-js";
import path from "path";
import * as fs from "node:fs";
import { parseArgs } from "node:util";
import { marked } from "marked";
import { getSupabase } from "./utils";
import { resolvePathWithinRepository } from "./path-security";

function getAllMarkdownFiles(dirPath: string, base: string = ""): string[] {
  let files: string[] = [];
  const entries = fs.readdirSync(dirPath, { withFileTypes: true });

  for (const entry of entries) {
    const relPath = base ? path.join(base, entry.name) : entry.name;
    const fullPath = path.join(dirPath, entry.name);

    if (entry.isDirectory()) {
      files = files.concat(getAllMarkdownFiles(fullPath, relPath));
    } else if (entry.isFile() && path.extname(entry.name) === ".md") {
      files.push(relPath.split(path.sep).join("/"));
    }
  }

  return files;
}

function markdownToHtml(mdText: string): string {
  return marked.parse(mdText) as string;
}

async function syncTexts(
  supabase: SupabaseClient<Database>,
  markdownFilesPath: string,
): Promise<void> {
  const baseDir = resolvePathWithinRepository(markdownFilesPath);
  const markdownFiles = getAllMarkdownFiles(baseDir);

  if (markdownFiles.length === 0) {
    console.log(
      `⚠️  No Markdown files found in ${baseDir}. Nothing to upload.`,
    );
    return;
  }

  // Clear existing records where id != 0
  const { error: deleteError } = await supabase
    .from("phase_texts")
    .delete()
    .neq("id", 0);

  if (deleteError) {
    console.error(
      `❌ Failed to clear existing records: ${deleteError.message}`,
    );
    return;
  }

  for (const relPath of markdownFiles) {
    const fullPath = path.join(baseDir, relPath);
    const mdContent = fs.readFileSync(fullPath, "utf-8");
    const htmlContent = markdownToHtml(mdContent);

    try {
      const { data, error } = await supabase
        .from("phase_texts")
        .upsert({ path: relPath, html_content: htmlContent })
        .select();

      if (error) throw error;

      console.log(`✅ Uploaded ${relPath} (${data ? data.length : 0} row(s))`);
    } catch (e: unknown) {
      const message = e instanceof Error ? e.message : String(e);
      console.error(`❌ Failed to upload ${relPath}: ${message}`);
    }
  }

  console.log("✅ All markdown files uploaded to Supabase");
}

(async () => {
  const supabase = getSupabase();

  const { values, positionals } = parseArgs({
    options: {
      file: {
        type: "string",
        short: "f",
      },
    },
    allowPositionals: true,
  });

  const filePath = values.file ?? positionals[0];

  if (!filePath) {
    console.error(
      "Error: Please provide a file via '-f <file>' or as the first argument.",
    );
    return;
  }

  await syncTexts(supabase, filePath);
})().catch((e) => console.error(e));
